src/de.zig

```diff
@@ -23,39 +23,55 @@ pub const de = struct {
         /// A deserialization block.
         comptime user_dbt: anytype,
     ) void {
-        return getty.de.free(allocator, Deserializer(user_dbt).@"getty.Deserializer", value);
+        return getty.de.free(
+            allocator,
+            Deserializer(
+                user_dbt,
+                // TODO: wonk
+                std.io.FixedBufferStream([]u8).Reader,
+            ).@"getty.Deserializer",
+            value,
+        );
     }
 };
 
 /// Deserializes into a value of type `T` from the deserializer `d`.
 pub fn fromDeserializer(comptime T: type, d: anytype) !T {
     const value = try getty.deserialize(d.allocator, T, d.deserializer());
-    errdefer if (d.allocator) |alloc| de.free(alloc, value, null);
+    errdefer de.free(d.allocator, value, null);
     try d.end();
 
     return value;
 }
 
 /// Deserializes into a value of type `T` from a slice of JSON using a deserialization block or tuple.
-pub fn fromSliceWith(allocator: ?std.mem.Allocator, comptime T: type, slice: []const u8, comptime user_dbt: anytype) !T {
-    const D = Deserializer(user_dbt);
-    var d = if (allocator) |alloc| D.withAllocator(alloc, slice) else D.init(slice);
-
-    return fromDeserializer(T, &d);
+pub fn fromSliceWith(allocator: std.mem.Allocator, comptime T: type, slice: []const u8, comptime user_dbt: anytype) !T {
+    var fbs = std.io.fixedBufferStream(slice);
+    return try fromReaderWith(allocator, T, fbs.reader(), user_dbt);
 }
 
 /// Deserializes into a value of type `T` from a slice of JSON.
-pub fn fromSlice(allocator: ?std.mem.Allocator, comptime T: type, slice: []const u8) !T {
+pub fn fromSlice(allocator: std.mem.Allocator, comptime T: type, slice: []const u8) !T {
     return try fromSliceWith(allocator, T, slice, null);
 }
 
+pub fn fromReader(allocator: std.mem.Allocator, comptime T: type, reader: anytype) !T {
+    return try fromReaderWith(allocator, T, reader, null);
+}
+
+pub fn fromReaderWith(allocator: std.mem.Allocator, comptime T: type, reader: anytype, comptime user_dbt: anytype) !T {
+    var d = Deserializer(user_dbt, @TypeOf(reader)).init(allocator, reader);
+    defer d.deinit();
+    return fromDeserializer(T, &d);
+}
```

Comment:
Can you move these function above `fromSliceWith` and swap their positions?

Reading from the top of the file, we should get the innermost function `fromDeserializer`, then `fromReaderWith`, which uses `fromDeserializer`, then `fromReader`, which uses `fromReaderWith`, and so on.